secret-sync-controller/controllers/secretsync_controller.go

```diff
@@ -0,0 +1,496 @@
+		prevSecretHash := ss.Status.SyncHash
+		prevTime := ss.Status.LastSuccessfulSyncTime
+
+		// Update status fields.
+		ss.Status.LastSuccessfulSyncTime = &metav1.Time{Time: time.Now()}
+		ss.Status.SyncHash = syncHash
+
+		if failedCondition.Type != "" {
+			meta.RemoveStatusCondition(&ss.Status.Conditions, failedCondition.Type)
+		}
+
+		// Attempt to create or update the secret.
+		if err = r.serverSidePatchSecret(ctx, ss, secretName, req.Namespace, datamap, objectVersions, labels, annotations, secretType); err != nil {
+			logger.Error(err, "failed to patch secret", "secretName", secretName)
+
+			// Rollback to the previous hash and the previous last successful sync time.
+			ss.Status.SyncHash = prevSecretHash
+			ss.Status.LastSuccessfulSyncTime = prevTime
+
+			// Reset the create or update conditions
+			meta.RemoveStatusCondition(&ss.Status.Conditions, ConditionTypeCreate)
+			meta.RemoveStatusCondition(&ss.Status.Conditions, ConditionTypeUpdate)
+
+			if checkIfErrorMessageCanBeDisplayed(err.Error()) {
+				failedCondition.Message = err.Error()
+				r.updateStatusConditions(ctx, ss, ConditionTypeUnknown, conditionType, ConditionReasonValidatingAdmissionPolicyCheckFailed, true)
+			} else {
+				r.updateStatusConditions(ctx, ss, ConditionTypeUnknown, conditionType, ConditionReasonSecretPatchFailedUnknownError, true)
+			}
+
+			return ctrl.Result{}, err
+		}
+
+		// No errors found, remove the failed conditions.
+		for _, cond := range ss.Status.Conditions {
+			if slices.Contains(FailedConditionsTriggeringRetry, cond.Reason) {
+				meta.RemoveStatusCondition(&ss.Status.Conditions, cond.Type)
+			}
+		}
+	}
+
+	// Update the status.
+	err = r.Client.Status().Update(ctx, ss)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	logger.V(4).Info("Done... updated status", "syncHash", syncHash, "lastSuccessfulSyncTime", ss.Status.LastSuccessfulSyncTime)
+	return ctrl.Result{}, nil
+}
+
+// checkIfErrorMessageCanBeDisplayed checks if the error message can be displayed in the condition message
+// based on the allowed strings to display condition error message defined in the conditions.go file.
+func checkIfErrorMessageCanBeDisplayed(errorMessage string) bool {
+	for _, allowedString := range AllowedStringsToDisplayConditionErrorMessage {
+		if strings.Contains(strings.ToLower(errorMessage), allowedString) {
+			return true
+		}
+	}
+	return false
+}
+
+// serverSidePatchSecret performs a server-side patch on a Kubernetes Secret.
+// It updates the specified secret with the provided data, labels, and annotations.
+func (r *SecretSyncReconciler) serverSidePatchSecret(ctx context.Context, ss *secretsyncv1alpha1.SecretSync, name, namespace string, datamap map[string][]byte, _, labels, annotations map[string]string, secretType corev1.SecretType) (err error) {
+	logger := log.FromContext(ctx)
+
+	secretKind := "Secret"
+	secretVersion := "v1"
+
+	// Retrieve Kubernetes clientset.
+	clientset := kubernetes.NewForConfigOrDie(ctrl.GetConfigOrDie())
```

Comment:
This should be stored as part of the controller struct instead of initializing here every time